src/java.base/share/classes/jdk/internal/util/ArraysSupport.java

```diff
@@ -575,52 +575,83 @@ public static int mismatch(double[] a, int aFromIndex,
+     * minimum growth amount is used as the preferred growth amount.
+     *
+     * The preferred length is determined by adding the preferred growth amount to the
+     * current length. If the preferred length does not exceed the soft maximum length
+     * (SOFT_MAX_ARRAY_LENGTH) then the preferred length is returned.
      *
-     * @param oldLength   current length of the array (must be non negative)
-     * @param minGrowth   minimum required growth of the array length (must be
-     *                    positive)
-     * @param prefGrowth  preferred growth of the array length (ignored, if less
-     *                    then {@code minGrowth})
+     * Since the preferred length exceeds the soft maximum, we use the minimum growth
+     * amount. The minimum required length is determined by adding the minimum growth
+     * amount to the current length. If the minimum required length is less than the
+     * soft maximum, the soft maximum is returned. If the minimum required length is
+     * greater than the soft maximum but does not exceed Integer.MAX_VALUE, the minimum
+     * required length is returned. Otherwise, the minimum required length exceeds
+     * Integer.MAX_VALUE, which can never be fulfilled, so this method throws OutOfMemoryError.
+     *
+     * Note that this method does not do any array allocation itself; it only does array
+     * length growth computations. However, it will throw OutOfMemoryError as noted above.
+     *
+     * Note also that this method cannot detect the JVM's implementation limit, and it
+     * may compute and return a length value up to and including Integer.MAX_VALUE that
+     * might exceed the JVM's implementation limit. In that case, the caller will likely
+     * attempt an array allocation with that length and encounter an OutOfMemoryError.
+     * Of course, regardless of the length value returned from this method, the caller
+     * may encounter OutOfMemoryError if there is insufficient heap to fufill the request.
+     *
+     * @param oldLength   current length of the array (must be nonnegative)
+     * @param minGrowth   minimum required growth amount (must be positive)
+     * @param prefGrowth  preferred growth amount
      * @return the new length of the array
-     * @throws OutOfMemoryError if increasing {@code oldLength} by
-     *                    {@code minGrowth} overflows.
+     * @throws OutOfMemoryError if the new length would exceed Integer.MAX_VALUE
      */
     public static int newLength(int oldLength, int minGrowth, int prefGrowth) {
         // assert oldLength >= 0
         // assert minGrowth > 0
 
-        int newLength = Math.max(minGrowth, prefGrowth) + oldLength;
-        if (newLength - MAX_ARRAY_LENGTH <= 0) {
-            return newLength;
+        int prefLength = oldLength + Math.max(minGrowth, prefGrowth); // might overflow
+        if (0 < prefLength && prefLength <= SOFT_MAX_ARRAY_LENGTH) {
+            return prefLength;
```

Comment:
The preconditions aren't checked, because this is an internal method, and the code size is minimized in order to help inlining. That's also why `hugeLength` is a separate method. (I guess I could add comments to this effect.) With this in mind it's hard to reason about robustness. If prefLength is zero, this can only be because some precondition was violated (e.g., oldLength is negative). If this were to occur there doesn't seem to be any advantage choosing one undefined behavior over another.

Reply:
Is there a reason the code would not naturally work when either min or preferred is zero?
Why are their preconditions?  What do they allow?

Reply:
The origin of this code is in collections like ArrayList that have an existing array (hence oldLength >= 0) and that need it to grow (hence minGrowth > 0). Those were the prevailing assumptions in the code from which this was derived, so they turn into preconditions here. I don't see the need to try to make this handle any more cases than it currently does. Doing so complicates the analysis and possibly the code as well. Certainly a bug in a caller might be difficult to track down, but I don't want to add argument checking or to provide "reasonable" behavior in the face of unreasonable inputs. This is an internal method; bugs in callers should be fixed.

Reply:
The algorithm can be well defined for minGrowth and prefGrowth == 0 without extra checks or exceptions with a careful look at the inequality checks.

Reply:
That's only true if oldLength is zero. The behavior is different if oldLength is positive. In any case, this method is called when the array needs to **grow**, which means the caller needs to reallocate and copy. If the caller passes zero for both minGrowth and prefGrowth, the caller doesn't need to reallocate and copy, and there's no point in it calling this method in the first place.

Reply:
Not calling for a zero value is usually an optimization, not a necessity to work around an inconsistency or unpredictability in the API or the range of arguments it accepts.